lib/src/socket_connection.dart

```diff
@@ -0,0 +1,444 @@
+
+import 'package:logging/logging.dart';
+import 'package:phoenix_socket/phoenix_socket.dart';
+import 'package:phoenix_socket/src/delayed_callback.dart';
+import 'package:rxdart/rxdart.dart';
+import 'package:web_socket_channel/status.dart';
+import 'package:web_socket_channel/web_socket_channel.dart';
+
+part 'socket_state.dart';
+
+typedef WebSocketChannelFactory = Future<WebSocketChannel> Function();
+
+final _logger = Logger('phoenix_socket.connection');
+
+// Some custom close codes.
+const heartbeatTimedOut = 4001;
+const forcedReconnectionRequested = 4002;
+
+/// Maintains connection to the underlying websocket, reconnecting to it if
+/// necessary.
+class SocketConnectionManager {
+  SocketConnectionManager({
+    required WebSocketChannelFactory factory,
+    required List<Duration> reconnectDelays,
+    required void Function(String) onMessage,
+    required void Function(WebSocketConnectionState) onStateChange,
+    required void Function(Object, [StackTrace?]) onError,
+  })  : _factory = factory,
+        _reconnectDelays = reconnectDelays,
+        _onError = onError,
+        _onStateChange = onStateChange,
+        _onMessage = onMessage;
+
+  final WebSocketChannelFactory _factory;
+  final List<Duration> _reconnectDelays;
+  final void Function(String message) _onMessage;
+  final void Function(WebSocketConnectionState state) _onStateChange;
+  final void Function(Object error, [StackTrace? stacktrace]) _onError;
+
+  /// Currently attempted or live connection. Value is null only before
+  /// any connection was attempted.
+  final _connectionsStream = BehaviorSubject<_WebSocketConnection>();
+
+  /// Count of consecutive attempts at establishing a connection without
+  /// success.
+  int _connectionAttempts = 0;
+
+  /// References to the last initialized connection attempt.
+  DelayedCallback<_WebSocketConnection>? _currentAttempt;
+
+  bool _disposed = false;
+
+  /// Requests to start connecting to the socket. Returns a future that
+  /// completes when a WebSocket connection has been established.
+  ///
+  /// If [immediately] is false, and a connection is already established, then
+  /// this method does not have any effect.
+  ///
+  /// If [immediately] is set to true, then an attempt to connect is made
+  /// immediately. This might result in dropping the current connection and
+  /// establishing a new one.
+  Future<void> start({bool immediately = false}) async {
+    if (_disposed) {
+      throw StateError('Cannot start: WebSocket connection manager disposed');
+    }
+
+    if (immediately) {
+      final currentAttempt = _currentAttempt;
+      if (_connectionsStream.hasValue &&
+          currentAttempt != null &&
+          !currentAttempt.delayDone) {
+        currentAttempt.skipDelay();
+        return;
+      }
+
+      reconnect(
+        forcedReconnectionRequested,
+        reason: 'Immediate connection requested',
+        immediately: true,
+      );
+    }
+    await _maybeConnect();
```

Comment:
Is it safe to call `_maybeConnect()` if `immediately == true`?
`reconnect()` also calls `_connect()`

Reply:
It is safe, but just to be clear made it so that we return inside the `if`.